stl/inc/xstring

```diff
@@ -2526,38 +2519,33 @@ private:
         _Apply_annotation(_My_data._Myptr(), _My_data._Myres, _My_data._Mysize, _My_data._Myres);
     }
 
-    _CONSTEXPR20 void _Modify_annotation(const difference_type _Count) const noexcept {
-        // Extends/shrinks the annotated range by _Count
-        if (_Count == 0) {
-            return;
-        }
-
+    _CONSTEXPR20 void _Modify_annotation(const size_type _Old_size, const size_type _New_size) const noexcept {
         auto& _My_data = _Mypair._Myval2;
-        _Apply_annotation(
-            _My_data._Myptr(), _My_data._Myres, _My_data._Mysize, static_cast<size_type>(_My_data._Mysize + _Count));
+        _Apply_annotation(_My_data._Myptr(), _My_data._Myres, _Old_size, _New_size);
     }
 
-    _NODISCARD static const void* _Get_aligned_first(const void* _First, const size_type _Capacity) noexcept {
-        const char* _CFirst = reinterpret_cast<const char*>(_First);
+    _CONSTEXPR20 void _Annotate_switch_to_large() const noexcept {
+        // annotates exactly the pointer as valid
+        _Remove_annotation();
 
-        if (_Capacity >= _Asan_granularity) { // We are guaranteed to have sufficient space to find an aligned address
-            return reinterpret_cast<const void*>(
-                (reinterpret_cast<uintptr_t>(_CFirst) + (_Asan_granularity - 1)) & ~(_Asan_granularity - 1));
-        }
+        auto& _My_data              = _Mypair._Myval2;
+        const void* const _My_buf   = _My_data._Bx._Buf;
+        const void* const _Ptr_last = _STD addressof(_My_data._Bx._Ptr) + 1;
+        const void* const _Buf_last = _My_data._Bx._Buf + _BUF_SIZE;
 
-        uintptr_t _Alignment_offset = reinterpret_cast<uintptr_t>(_CFirst) & (_Asan_granularity - 1);
-        if (_Alignment_offset != 0) {
-            _Alignment_offset = _Asan_granularity - _Alignment_offset;
-        }
+        if constexpr (_Small_string_always_aligned) {
+            __sanitizer_annotate_contiguous_container(_My_buf, _Buf_last, _Buf_last, _Ptr_last);

```

Comment:
Does this depend on `sizeof(_My_data._Bx._Ptr)` being a multiple of 8, or 4, or anything? I've forgotten half of what I once understood about ASan. (Also on 2541.)

Reply:
it does not. The way it works is that each 8-byte region can be marked "first-N-valid", so having 1 byte, 2 bytes, etc. is fine as long as the first byte is on an asan boundary.